README.md

# diyar

[![CRAN
version](http://www.r-pkg.org/badges/version/diyar)](https://cran.r-project.org/package=diyar)
[![CRAN RStudio mirror
downloads](https://cranlogs.r-pkg.org/badges/diyar)](https://www.r-pkg.org/pkg/diyar)
[![Coverage
status](https://codecov.io/gh/OlisaNsonwu/diyar/branch/master/graph/badge.svg)](https://app.codecov.io/github/OlisaNsonwu/diyar?branch=master)
[![Travis build
status](https://api.travis-ci.com/OlisaNsonwu/diyar.svg?branch=master)](https://app.travis-ci.com/github/OlisaNsonwu/diyar)

## Installation

``` r
# Install the latest CRAN release 
install.packages("diyar")

# Or, install the development version from GitHub
install.packages("devtools")
devtools::install_github("OlisaNsonwu/diyar")
```

## Overview

`diyar` is an `R` package for linking records with shared
characteristics. The linked records represent an entity, which depending
on the context of the analysis can be unique patients, infection
episodes, overlapping periods of care, clusters or other occurrences as
defined by a case definition. This makes it useful in ordinarily complex
analyses such as record linkage,  
contact or network analyses e.t.c.

The main functions are `links()`, `episodes()` and `partitions()`. They
are flexible in regards to how they compare records, as well as what are
considered matches. Their functionality can sometimes overlap however,
each is better suited to particular use cases:

- `links()` - link records with no relevance to an index record. For
  example, deterministic record linkage
- `episodes()` - link records in relation to an index record. For
  example, contact and network analysis.  
- `partitions()` - link records in relation to a fixed interval.

## links()

**Key features;**

- multi-stage record linkage. Here, multiple linkage criteria are
  assessed in a specified order of priority.

``` r
library(diyar)
data(missing_staff_id)
dfr_stages <- missing_staff_id[c("age", "hair_colour", "branch_office")]
priority_order_1 <- c("hair_colour", "branch_office")
priority_order_2 <- c("branch_office", "hair_colour")

dfr_stages$id.1 <- links(criteria = as.list(dfr_stages[priority_order_1]))
dfr_stages$id.2 <- links(criteria = as.list(dfr_stages[priority_order_2]))
```

- create and use complex rules for record matching. This is done with a
  `sub_criteria()`.

``` r
sub.cri.1 <- sub_criteria(
  hair.color = dfr_stages$hair_colour,
  age = dfr_stages$age,
  match_funcs = c(
    "exact" = exact_match,
    "age.range" = range_match)
)
last_word_wf <- function(x) tolower(gsub("^.* ", "", x))
last_word_cmp <- function(x, y) last_word_wf(x) == last_word_wf(y)
not_equal <- function(x, y) x != y
sub.cri.2 <- sub_criteria(
  dfr_stages$branch_office, 
  dfr_stages$age,
  match_funcs = c(
    "last.word" = last_word_cmp,
    "not.equal" = not_equal)
)
sub.cri.3 <- sub_criteria(sub.cri.1, sub.cri.2, operator = "and")
sub.cri.1
#> {
#> exact(hair.color) OR age.range(age)
#> }
sub.cri.2
#> {
#> last.word(Republic of Ghana,France,NA ...) OR not.equal(30,30,30 ...)
#> }
sub.cri.3
#> {
#>   {
#>   exact(hair.color) OR age.range(age)
#>   } AND 
#>   {
#>   last.word(Republic of Ghana,France,NA ...) OR not.equal(30,30,30 ...)
#>   }
#> }
dfr_stages$id.3 <- links(
  criteria = "place_holder",
  sub_criteria = list("cr1" = sub.cri.3)
)
dfr_stages
#>   age hair_colour     branch_office          id.1          id.2          id.3
#> 1  30       Brown Republic of Ghana P.1 (CRI 001) P.1 (CRI 001) P.1 (CRI 001)
#> 2  30        Teal            France P.4 (CRI 003) P.2 (CRI 001) P.2 (CRI 001)
#> 3  30        <NA>              <NA> P.3 (No hits) P.3 (No hits) P.3 (No hits)
#> 4  30       Green              <NA> P.4 (CRI 001) P.2 (CRI 003) P.4 (No hits)
#> 5  30       Green            France P.4 (CRI 001) P.2 (CRI 001) P.2 (CRI 001)
#> 6  30  Dark brown             Ghana P.6 (No hits) P.6 (No hits) P.1 (CRI 001)
#> 7  30       Brown Republic of Ghana P.1 (CRI 001) P.1 (CRI 001) P.1 (CRI 001)
```

There are variations of `links()` like `links_wf_probabilistic()` and
`links_af_probabilistic()` for specific use cases such as probabilistic
record linkage.

## episodes()

**Key features;**

- link records within a specified period from an index record.

``` r
dfr_2 <- data.frame(date = as.Date("2020-01-01") + c(1:5, 10:15, 20:25))
dfr_2$id.1 <- episodes(
  date = dfr_2$date, case_length = 2,
  episodes_max = 1)
```

- change the index record.

``` r
dfr_2$pref <- c(rep(2, 8), 1, rep(2, 8))
dfr_2$id.2 <- episodes(
  date = dfr_2$date, case_length = number_line(-2, 2),
  episodes_max = 1, 
  custom_sort = dfr_2$pref)
```

- add a recurrence period

``` r
dfr_2$id.3 <- episodes(
  date = dfr_2$date, case_length = number_line(-2, 2), 
  episode_type = "rolling", recurrence_length = 1,
  episodes_max = 1, rolls_max = 1)
```

- link overlapping periods

``` r
dfr_2$period <- number_line(dfr_2$date, dfr_2$date + 5)
dfr_2$id.4 <- episodes(
  date = dfr_2$period, case_length = index_window(dfr_2$period),
  episodes_max = 1)
dfr_2
#>          date                              id.1 pref
#> 1  2020-01-02 E.01 2020-01-02 -> 2020-01-04 (C)    2
#> 2  2020-01-03 E.01 2020-01-02 -> 2020-01-04 (D)    2
#> 3  2020-01-04 E.01 2020-01-02 -> 2020-01-04 (D)    2
#> 4  2020-01-05 E.04 2020-01-05 == 2020-01-05 (S)    2
#> 5  2020-01-06 E.05 2020-01-06 == 2020-01-06 (S)    2
#> 6  2020-01-11 E.06 2020-01-11 == 2020-01-11 (S)    2
#> 7  2020-01-12 E.07 2020-01-12 == 2020-01-12 (S)    2
#> 8  2020-01-13 E.08 2020-01-13 == 2020-01-13 (S)    2
#> 9  2020-01-14 E.09 2020-01-14 == 2020-01-14 (S)    1
#> 10 2020-01-15 E.10 2020-01-15 == 2020-01-15 (S)    2
#> 11 2020-01-16 E.11 2020-01-16 == 2020-01-16 (S)    2
#> 12 2020-01-21 E.12 2020-01-21 == 2020-01-21 (S)    2
#> 13 2020-01-22 E.13 2020-01-22 == 2020-01-22 (S)    2
#> 14 2020-01-23 E.14 2020-01-23 == 2020-01-23 (S)    2
#> 15 2020-01-24 E.15 2020-01-24 == 2020-01-24 (S)    2
#> 16 2020-01-25 E.16 2020-01-25 == 2020-01-25 (S)    2
#> 17 2020-01-26 E.17 2020-01-26 == 2020-01-26 (S)    2
#>                                 id.2                              id.3
#> 1  E.01 2020-01-02 == 2020-01-02 (S) E.01 2020-01-02 -> 2020-01-05 (C)
#> 2  E.02 2020-01-03 == 2020-01-03 (S) E.01 2020-01-02 -> 2020-01-05 (D)
#> 3  E.03 2020-01-04 == 2020-01-04 (S) E.01 2020-01-02 -> 2020-01-05 (D)
#> 4  E.04 2020-01-05 == 2020-01-05 (S) E.01 2020-01-02 -> 2020-01-05 (R)
#> 5  E.05 2020-01-06 == 2020-01-06 (S) E.05 2020-01-06 == 2020-01-06 (S)
#> 6  E.06 2020-01-11 == 2020-01-11 (S) E.06 2020-01-11 == 2020-01-11 (S)
#> 7  E.09 2020-01-12 -> 2020-01-16 (D) E.07 2020-01-12 == 2020-01-12 (S)
#> 8  E.09 2020-01-12 -> 2020-01-16 (D) E.08 2020-01-13 == 2020-01-13 (S)
#> 9  E.09 2020-01-12 -> 2020-01-16 (C) E.09 2020-01-14 == 2020-01-14 (S)
#> 10 E.09 2020-01-12 -> 2020-01-16 (D) E.10 2020-01-15 == 2020-01-15 (S)
#> 11 E.09 2020-01-12 -> 2020-01-16 (D) E.11 2020-01-16 == 2020-01-16 (S)
#> 12 E.12 2020-01-21 == 2020-01-21 (S) E.12 2020-01-21 == 2020-01-21 (S)
#> 13 E.13 2020-01-22 == 2020-01-22 (S) E.13 2020-01-22 == 2020-01-22 (S)
#> 14 E.14 2020-01-23 == 2020-01-23 (S) E.14 2020-01-23 == 2020-01-23 (S)
#> 15 E.15 2020-01-24 == 2020-01-24 (S) E.15 2020-01-24 == 2020-01-24 (S)
#> 16 E.16 2020-01-25 == 2020-01-25 (S) E.16 2020-01-25 == 2020-01-25 (S)
#> 17 E.17 2020-01-26 == 2020-01-26 (S) E.17 2020-01-26 == 2020-01-26 (S)
#>                      period                              id.4
#> 1  2020-01-02 -> 2020-01-07 E.01 2020-01-02 -> 2020-01-11 (C)
#> 2  2020-01-03 -> 2020-01-08 E.01 2020-01-02 -> 2020-01-11 (D)
#> 3  2020-01-04 -> 2020-01-09 E.01 2020-01-02 -> 2020-01-11 (D)
#> 4  2020-01-05 -> 2020-01-10 E.01 2020-01-02 -> 2020-01-11 (D)
#> 5  2020-01-06 -> 2020-01-11 E.01 2020-01-02 -> 2020-01-11 (D)
#> 6  2020-01-11 -> 2020-01-16 E.06 2020-01-11 -> 2020-01-16 (S)
#> 7  2020-01-12 -> 2020-01-17 E.07 2020-01-12 -> 2020-01-17 (S)
#> 8  2020-01-13 -> 2020-01-18 E.08 2020-01-13 -> 2020-01-18 (S)
#> 9  2020-01-14 -> 2020-01-19 E.09 2020-01-14 -> 2020-01-19 (S)
#> 10 2020-01-15 -> 2020-01-20 E.10 2020-01-15 -> 2020-01-20 (S)
#> 11 2020-01-16 -> 2020-01-21 E.11 2020-01-16 -> 2020-01-21 (S)
#> 12 2020-01-21 -> 2020-01-26 E.12 2020-01-21 -> 2020-01-26 (S)
#> 13 2020-01-22 -> 2020-01-27 E.13 2020-01-22 -> 2020-01-27 (S)
#> 14 2020-01-23 -> 2020-01-28 E.14 2020-01-23 -> 2020-01-28 (S)
#> 15 2020-01-24 -> 2020-01-29 E.15 2020-01-24 -> 2020-01-29 (S)
#> 16 2020-01-25 -> 2020-01-30 E.16 2020-01-25 -> 2020-01-30 (S)
#> 17 2020-01-26 -> 2020-01-31 E.17 2020-01-26 -> 2020-01-31 (S)
```

There are variations of `episodes()` like `episodes_wf_splits()` for
specific use cases such as more efficient handling of duplicate records.

## partitions()

**Key features;**

- link all records within a specific periods in time

``` r
dfr_3 <- dfr_2["date"]
dfr_3$id.1 <- partitions(
  date = dfr_3$date, 
  window = number_line(as.Date(c("2020-01-10", "2020-01-17")), 
                       as.Date(c("2020-01-12", "2020-01-24")))
  )
```

- link all records within a splits of an interval

``` r
dfr_3$id.2 <- partitions(date = dfr_3$date, by = 3, separate = TRUE) 
dfr_3$id.3 <- partitions(date = dfr_3$date, length.out = 3, separate = TRUE)
dfr_3
#>          date      id.1      id.2      id.3
#> 1  2020-01-02 PN.01 (I) PN.18 (I) PN.18 (I)
#> 2  2020-01-03 PN.02 (I) PN.18 (D) PN.18 (D)
#> 3  2020-01-04 PN.03 (I) PN.18 (D) PN.18 (D)
#> 4  2020-01-05 PN.04 (I) PN.19 (I) PN.18 (D)
#> 5  2020-01-06 PN.05 (I) PN.19 (D) PN.18 (D)
#> 6  2020-01-11 PN.06 (I) PN.21 (I) PN.18 (D)
#> 7  2020-01-12 PN.07 (I) PN.21 (D) PN.18 (D)
#> 8  2020-01-13 PN.08 (I) PN.21 (D) PN.18 (D)
#> 9  2020-01-14 PN.09 (I) PN.22 (I) PN.19 (I)
#> 10 2020-01-15 PN.10 (I) PN.22 (D) PN.19 (D)
#> 11 2020-01-16 PN.11 (I) PN.22 (D) PN.19 (D)
#> 12 2020-01-21 PN.12 (I) PN.24 (I) PN.19 (D)
#> 13 2020-01-22 PN.13 (I) PN.24 (D) PN.19 (D)
#> 14 2020-01-23 PN.14 (I) PN.25 (I) PN.19 (D)
#> 15 2020-01-24 PN.15 (I) PN.25 (D) PN.19 (D)
#> 16 2020-01-25 PN.16 (I) PN.25 (D) PN.19 (D)
#> 17 2020-01-26 PN.17 (I) PN.25 (D) PN.19 (D)
```

Find out more!

- number_line and overlaps - `vignette("number_line")`
- Introduction to epidemiological case definitions with diyar -
  `vignette("episodes")`
- Introduction to record linkage with diyar - `vignette("links")`
- Divvy up events with partitions - `vignette("panes")`

## Bugs and issues

Please report any bug or issues with using this package
[here](https://github.com/OlisaNsonwu/diyar/issues).


NEWS.md

# Version 0.5.1

## New features

## Changes

## Bug fixes

- `links()` - Incorrect results in some situations. Resolved.
- `links_af_probabilistic()` - Failed in some situations. Resolved.

# Version 0.5.0

## New features

- New option (`"semi"`) for the `batched` argument in `links()`. All
  matches are compared against the record-set in the next iteration.
  Therefore, the number of record-pairs increase exponentially as new
  matches are found. This means fewer record-pairs (memory usage) but a
  longer run time compared to the `"no"` option. Conversely, it leads to
  more record-pairs (memory usage) but a shorter run time compared to
  the `"yes"` option.
- New argument (`batched`) in `episodes()`
- New argument (`split`) in `episodes()`. Split the analysis in
  `N`-splits of `strata`. This leads to fewer record-pairs (and memory
  usage) but a longer run time.
- New argument (`decode`) in `as.data.frame.pid()`,
  `as.data.frame.epid()` and `as.data.frame.pane()`
- New function - `episodes_af_shift()`. A more vectorised approach to
  `episodes()` based on `epidm::group_time()`.
- New function - `links_wf_episodes()`. Implantation of `episodes()`
  using `links()`.

## Changes

- Optimised `episodes()` and `links()`. Each iteration now uses less
  time and memory.
- `link_id` slot in `pid` objects is now a `list`.
- `links()` - records with missing values in a `sub_criteria` are now
  skipped at the corresponding iteration.
- Updated argument in `links()`- `recursive`. This now takes any of
  three options `[c("linked", "unlinked", "none")]` .
  `[c("linked", "unlinked")]` collectively were previously `[TRUE]`,
  while `["none"]` was previously `[FALSE]`.
- `as.epids()` now calls `make_episodes()`.
- The default value for the `window` argument in `partitions()` is now
  `NULL`
- `as.data.frame()` and `as.data.list()` now only creates
  elements/fields from non-empty fields
- `id` and `gid` slots in `number_line` objects are now `integer(0)` by
  default.
- `episode_group()`, `record_group()` and `range_match_legacy()` have
  been removed.
- `["recurisve"]` episodes from `episodes()` are now presented as
  `["rolling"]` episodes with `reference_event = "all_records"` i.e
  - `Old syntax ~ episodes(..., episode_type == "recursive")`
  - `New syntax ~ episodes(..., episode_type == "rolling", reference_event = "all_records")`

## Bug fixes

- When `recursive` was `TRUE`, `links()` ended prematurely and therefore
  missed some matches. Resolved.
- `recurrence_sub_criteria` in `episodes()` was not implemented
  correctly and lead to incorrect linkage result in some instances.
  Resolved.
- `overlap_method()` - logical tests recycled incorrectly. Resolved.
- `check_links` argument - Option `"g"` implemented as option `"l"`.
  Resolved.
- `make_pairs_wf_source()`. Created incorrect pairs. Resolved.
- `case_sub_criteria` and `recurrence_sub_criteria` in `episodes()` led
  to incorrect results. Resolved.

# Version 0.4.2

## New features

- New argument in `merge_ids()` - `shrink` and `expand`.
- New S3 method for class ‘d_report’ - `plot`.
- New S3 method for class ‘sub_criteria’ - `format`.
- New function - `true()`. Predefined logical test for use with
  `sub_criteria()`.
- New function - `false()`. Predefined logical test for use with
  `sub_criteria()`.
- New argument in `links()`- `batched`. Specify if all record pairs are
  created or compared at once (`"no"`) or in batches (`"yes"`).  
- New argument in `links()`- `repeats_allowed`. Specify if record-pairs
  with duplicate elements should be created.
- New argument in `links()`- `permutations_allowed`. Specify if
  permutations of the same record-pair should be created.
- New argument in `links()`- `ignore_same_source`. Specify if
  record-pairs from different datasets should be created.
  <!-- + New argument in `links_wf_probabilistic()`- `return_weights`. XXXXXX. -->
- New argument in `eval_sub_criteria()`- `depth`. First order of
  recursion.
- New function - `sets()` and `make_sets()`. Create permutations of
  record-sets.

## Changes

- `links()` - When `shrink` is `TRUE`, records in a record-group must
  meet every listed match `criteria` and `sub_criteria`. For example, if
  `pid_cri` is 3, then the record must have meet matched another on the
  the first three match criteria.
- `links()` - `pid@iteration` now tracks when a record was dealt with
  instead of when it was assigned to a record-group. For example, a
  record can be closed (matched or not matched) at iteration 1 but
  assigned to a record-group at iteration 5.
- `make_pairs()` - `x.*` and `y.*` values in the output are now swapped.
- `sub_criteria` can now export any data created by `match_func`. To do
  this, `match_func` must export a `list`, where the first element is a
  logical object. See an example below.

``` r
library(diyar)
val <- rep(month.abb[1:5], 2); val
#>  [1] "Jan" "Feb" "Mar" "Apr" "May" "Jan" "Feb" "Mar" "Apr" "May"
match_and_export <- function(x, y){
  output <- list(x == y, 
                 data.frame(x_val = x, y_val = y, is_match = x == y))
  return(output)
}
sub.cri.1 <- sub_criteria(
  val, match_funcs = list(match.export = match_and_export)
)

format(sub.cri.1, show_levels = TRUE)
#> logical_test-{
#> Lv.0.1-match.export(Jan,Feb,Mar ...)
#> }
eval_sub_criteria(sub.cri.1)
#> $logical_test
#>  [1] 1 0 0 0 0 1 0 0 0 0
#> 
#> $mf.0.1
#>    x_val y_val is_match
#> 1    Jan   Jan     TRUE
#> 2    Feb   Jan    FALSE
#> 3    Mar   Jan    FALSE
#> 4    Apr   Jan    FALSE
#> 5    May   Jan    FALSE
#> 6    Jan   Jan     TRUE
#> 7    Feb   Jan    FALSE
#> 8    Mar   Jan    FALSE
#> 9    Apr   Jan    FALSE
#> 10   May   Jan    FALSE
```

- `links` can now export any data created within a `sub_criteria`. To do
  this, the `sub_criteria` must be created as described above. See an
  example below


R/panes.R
#' @name partitions
#' @title Distribute events specified intervals.
#'
#' @description Distribute events into groups defined by time or numerical intervals.
#' Each set of linked records are assigned a unique identifier with relevant group-level data.
#'
#' @param sn \code{[integer]}. Unique record identifier. Useful for creating familiar \code{\link[=pane-class]{pane}} identifiers.
#' @param strata \code{[atomic]}. Subsets of the dataset. Panes are created separately for each \code{strata}.
#' @param windows_total \code{[integer|\link{number_line}]}. Minimum number of matched \code{windows} required for a pane. See \code{details}
#' @param separate \code{[logical]}. If \code{TRUE}, events matched to different \code{windows} are not linked.
#' @param date \code{[date|datetime|integer|\link{number_line}]}. Event date or period.
#' @param window \code{[integer|\link{number_line}]}. Numeric or time intervals.
#' @param data_source \code{[character]}. Unique data source identifier. Adds the list of datasets in each pane to the \code{\link[=pane-class]{pane}}. Useful when the dataset has data from multiple sources.
#' @param custom_sort \code{[atomic]}. Preferred order for selecting \code{"index"} events.
#' @param group_stats \code{[logical]}. If \code{TRUE} (default), the returned \code{pane} object will include group specific information like panes start and end dates.
#' @param data_links \code{[list|character]}. A set of \code{data_sources} required in each \code{\link[=pane-class]{pane}}. A \code{\link[=pane-class]{pane}} without records from these \code{data_sources} will be unlinked. See \code{Details}.
#' @param by \code{[integer]}. Width of splits.
#' @param length.out \code{[integer]}. Number of splits.
#' @param fill \code{[logical]}. Retain (\code{TRUE}) or drop (\code{FALSE}) the remainder of an uneven split.
#' @param display \code{[character]}. The progress messages printed on screen. Options are; \code{"none"} (default) or \code{"stats"}.
#' @return
#'
#' @return \code{\link[=pane-class]{pane}}
#'
#' @seealso
#' \code{\link[=pane-class]{pane}}; \code{\link{number_line_sequence}}; \code{\link{episodes}}; \code{\link{links}}; \code{\link{overlaps}}; \code{\link{number_line}}; \code{\link{schema}}
#'
#' @details
#' Each assigned group is referred to as a \code{\link[=pane-class]{pane}} A \code{\link[=pane-class]{pane}} consists of events within a specific time or numerical intervals (\code{window}).
#'
#' Each \code{window} must cover a separate interval. Overlapping \code{windows} are merged before events are distributed into panes.
#' Events that occur over two \code{windows} are assigned to the last one listed.
#'
#' Alternatively, you can create \code{windows} by splitting a period into equal parts (\code{length.out}), or into a sequence of intervals with fixed widths (\code{by}).
#'
#' By default, the earliest event is taken as the \code{"Index"} event of the \code{\link[=pane-class]{pane}}.
#' An alternative can be chosen with \code{custom_sort}.
#'
#' \bold{\code{partitions()}} will categorise records into 3 types;
#' \itemize{
#' \item \code{"Index"} - Index event/record of the pane.
#' \item \code{"Duplicate_I"} - Duplicate of the \code{"Index"} record.
#' \item \code{"Skipped"} - Records that are not assigned to a pane.
#' }
#'
#' Every element in \code{data_links} must be named \code{"l"} (links) or \code{"g"} (groups).
#' Unnamed elements of \code{data_links} will be assumed to be \code{"l"}.
#' \itemize{
#' \item If named \code{"l"}, only groups with records from every listed \code{data_source} will be retained.
#' \item If named \code{"g"}, only groups with records from any listed \code{data_source} will be retained.
#' }
#'
#' \emph{\code{NA} values in \code{strata} excludes records from the partitioning tracking process}.
#'
#' See \code{vignette("episodes")} for more information.
#'
#' @examples
#' events <- c(30, 2, 11, 10, 100)
#' windows <- number_line(c(1, 9, 25), c(3, 12, 35))
#'
#' events
#' partitions(date = events, length.out = 3, separate = TRUE)
#' partitions(date = events, by = 10, separate = TRUE)
#' partitions(date = events, window = windows, separate = TRUE)
#' partitions(date = events, window = windows, separate = FALSE)
#' partitions(date = events, window = windows, separate = FALSE, windows_total = 4)
#'
#' @aliases partitions
#' @export
#'
partitions <- function(date, window = number_line(0, Inf), windows_total = 1, separate = FALSE, sn = NULL, strata = NULL,
                  data_links = "ANY", custom_sort = NULL, group_stats = FALSE,
                  data_source = NULL, by = NULL, length.out = NULL, fill =  TRUE, display = "none"){
  tm_a <- Sys.time()

  # Validations
  errs <- err_partitions_checks_0(date = date,
                             window = window,
                             windows_total = windows_total,
                             separate = separate,
                             sn = sn,
                             strata = strata,
                             data_source = data_source,
                             data_links = data_links,
                             custom_sort = custom_sort,
                             group_stats = group_stats,
                             by = by,
                             length.out = length.out,
                             fill = fill,
                             display = display
                             )

  if(errs != FALSE) stop(errs, call. = FALSE)

  options_lst = list(date = as.number_line(date),
                     strata = strata,
                     separate = separate)

  dl_lst <- unlist(data_links, use.names = FALSE)
  # Standardise inputs
    # `data_links`
  if(!all(class(data_links) == "list")){
    data_links <- list(l = data_links)
  }
  if(is.null(names(data_links))) names(data_links) <- rep("l", length(data_links))
  names(data_links) <- ifelse(names(data_links) == "", "l", names(data_links))
  # `date`
  int <- as.number_line(date)
  int@id <- seq_len(length(int))
  # `strata`
  if(length(strata) == 1) {
    cri <- rep(1L, length(int))
    cri_l <- rep(strata, length(int))
  }else if(is.null(strata)){
    cri <- rep(1, length(int))
  }else{
    cri <- match(strata, strata[!duplicated(strata)])
    cri_l <- strata
  }
  # `sn`
  if(is.null(sn)) {
    sn <- seq_len(length(int))
  }else{
    sn <- as.integer(sn)
  }
  # `custom_sort`
  if(!is.null(custom_sort)) {
    c_sort <- as.numeric(as.factor(custom_sort))
    if(length(c_sort) == 1) c_sort <- rep(c_sort, length(int))
  }else{
    c_sort <- rep(0L, length(int))
  }
  # `windows_total`
  if(is.number_line(windows_total)){
    windows_total[windows_total@.Data < 0] <- reverse_number_line(windows_total[windows_total@.Data < 0], "decreasing")
  }else{
    windows_total <- number_line(windows_total, Inf)
  }
  windows_total[windows_total@.Data < 0] <- reverse_number_line(windows_total[windows_total@.Data < 0], "decreasing")

  if(length(windows_total) == 1){
    windows_total <- rep(windows_total, length(int))
  }
  # Strata-specific `windows`
  if(is.number_line(window)){
    window <- list(compress_number_line(as.number_line(window), deduplicate = TRUE, collapse = TRUE))
  }else{
    window <- as.number_line(window)
  }
  if(!is.null(by) | !is.null(length.out) | length(window) > 1){
    split_cri <- cri
  }else{
    split_cri <- rep(1L, length(int))
  }
  splits <- split(int, split_cri)
  splits_sn <- split(int@id, split_cri)
  if(!is.null(by)){
    if (length(by) == 1) by <- rep(by, length(int))
    split_bys <- split(by, split_cri)
    if (length(fill) == 1) fill <- rep(fill, length(int))
    split_fills <- split(fill, split_cri)
    splits_func <- function(x, n, f) {
      n <- n[!duplicated(n)]
      f <- f[!duplicated(f)]
      number_line_sequence(number_line(min(x@start),
                                       max(x@start + x@.Data)),
                           by = n, fill = f)
    }
    splits_windows <- mapply(splits_func, splits, split_bys, split_fills, SIMPLIFY = FALSE)
  }else if(is.null(by) & !is.null(length.out)){
    if (length(length.out) == 1) length.out <- rep(length.out, length(int))
    split_lnts <- split(length.out, split_cri)
    if (length(fill) == 1) fill <- rep(fill, length(int))
    split_fills <- split(fill, split_cri)
    splits_func <- function(x, n, f) {
      n <- n[!duplicated(n)]
      f <- f[!duplicated(f)]
      number_line_sequence(number_line(min(x@start),
                                       max(x@start + x@.Data)),
                           length.out = n, fill = f)
    }
    splits_windows <- mapply(splits_func, splits, split_lnts, split_fills, SIMPLIFY = FALSE)
  }else{
    if(length(window) == 1){
      splits_windows <- window
      names(splits_windows) <- "1"
    }else{
      splits_windows <- split(window, split_cri)
      splits_windows <- lapply(splits_windows, function(x) x[!duplicated(x)])
      splits_windows <- unlist(splits_windows, recursive = FALSE)
      splits_windows <- lapply(splits_windows, function(x) compress_number_line(x, deduplicate = TRUE, collapse = TRUE))
    }
  }

  nms <- names(splits_windows)
  window_list <- splits_windows[match(split_cri, nms)]

  # Partition records into `windows`
  window_matched <- mapply(pane_checks, splits, splits_windows, SIMPLIFY = FALSE)
  window_matched <- unlist(window_matched, use.names = FALSE)
  splits_sn <- unlist(splits_sn, use.names = FALSE)
  window_matched <- window_matched[match(seq_len(length(int)), splits_sn)]

  case_nm <- ifelse(window_matched == 0, -1L, 1L)
  tag <- ifelse(window_matched == 0,
                NA_real_, cri + 1/window_matched)
  pane <- ifelse(!is.na(tag) & rep(!separate, length(int)),
                 cri, tag)
  # Implement `windows_total`
  dst <- rle(sort(cri[!is.na(tag) & !duplicated(tag)]))
  phits <- dst$lengths[match(cri, dst$values)]
  lgk <- !(phits >= as.numeric(windows_total@start) & phits <= as.numeric(right_point(windows_total))) | is.na(pane)
  pane[lgk] <- -seq_len(length(int))[lgk]

  # Index records - `custom_sort`
  ord <- order(pane, -c_sort, -as.numeric(int@start), -as.numeric(right_point(int)))
  s_pane <- pane[ord]
  s_sn <- sn[ord]
  # names(s_pane) <- s_sn
  # index_sn <- rle(s_pane)

  r <- rle(s_pane)
  lgk <- !duplicated(s_pane, fromLast = TRUE) | is.na(s_pane)
  pane <- rep(s_sn[lgk], r$lengths[match(s_pane[lgk], r$values)])
  pane <- pane[match(sn, s_sn)]
  case_nm[which(sn %in% s_sn[lgk] & case_nm != -1)] <- 0L
  pane[case_nm == -1] <- sn[case_nm == -1]

  # `pane_total`
  pane_n <- rep(r$lengths[match(s_pane[lgk], r$values)], r$lengths[match(s_pane[lgk], r$values)])
  pane_n <- pane_n[match(sn, s_sn)]

  # Distance of records from index record
  ei <- pane[which(case_nm %in% c(0, -1))]
  ii <- int[which(case_nm %in% c(0, -1))]
  lgk <- match(pane, ei)
  dist_pane_index <- ((as.numeric(int@start) + as.numeric(right_point(int))) * .5) -
    ((as.numeric(ii@start[lgk]) + as.numeric(right_point(ii[lgk]))) * .5)

  class(case_nm) <- "d_label"
  attr(case_nm, "value") <- c(-1, 0, 1)
  attr(case_nm, "label") <- c("Skipped", "Index", "Duplicate_I")
  attr(case_nm, "state") <- "encoded"
  # output - `pane` object
  panes <- new("pane",
               .Data = pane,
               dist_pane_index = dist_pane_index,
               window_matched = window_matched,
               sn = sn,
               case_nm = case_nm,
               window_list = window_list,
               pane_total = pane_n,
               options = options_lst)

  if(!is.null(data_source)){
    # Implement `data_links`
    rst <- check_links(panes, data_source, data_links)
    datasets <- rst$ds

    if(!all(toupper(dl_lst) == "ANY")){
      req_links <- rst$rq
      panes@dist_pane_index[!req_links] <- 0
      panes@case_nm[!req_links] <- -1L
      panes@.Data[!req_links] <- panes@sn[!req_links]
      # `pane_datasets`
      datasets[!req_links] <- data_source[!req_links]
    }
    panes@pane_dataset <- encode(datasets)
  }

  # `group_stats`
  if(group_stats == TRUE){
    lgk <- which(pane_n != 1)
    # `pane_interval`
    dts_a <- lapply(split(as.numeric(as.POSIXct(int@start[lgk]), tz = "GMT"), pane[lgk]), min)
    dts_z <- lapply(split(as.numeric(as.POSIXct(right_point(int[lgk])), tz = "GMT"), pane[lgk]), max)
    dts_a <- as.numeric(dts_a)[match(pane[lgk], names(dts_a))]
    dts_z <- as.numeric(dts_z)[match(pane[lgk], names(dts_z))]

    pane_dt_a <- as.numeric(int@start)
    pane_dt_z <- as.numeric(right_point(int))

    pane_dt_a[lgk] <- dts_a
    pane_dt_z[lgk] <- dts_z
    is_dt <- ifelse(!any(class(int@start) %in% c("Date","POSIXct","POSIXt","POSIXlt")), F, T)
    if(is_dt == TRUE){
      pane_dt_a <- as.POSIXct(pane_dt_a, "GMT", origin = as.POSIXct("1970-01-01", "GMT"))
      pane_dt_z <- as.POSIXct(pane_dt_z, "GMT", origin = as.POSIXct("1970-01-01", "GMT"))
      pane_l <- difftime(pane_dt_z, pane_dt_a, units = "days")
    }else{
      pane_l <- pane_dt_z - pane_dt_a
    }
    panes@pane_interval <- number_line(l = pane_dt_a, r = pane_dt_z, gid = pane)
    # `pane_length`
    panes@pane_length <- pane_l
  }

  tm_z <- Sys.time()
  tms <- difftime(tm_z, tm_a)
  tms <- paste0(ifelse(round(tms) == 0, "< 0.01", round(as.numeric(tms), 2)), " ", attr(tms, "units"))

  if(tolower(display) != "none") cat("Records partitioned in ", tms, "!\n", sep = "")
  return(panes)
}
